### DISCUSSION TITLE
GETting Culture Across APIs Homework

### DISCUSSION THREAD

**Discussion by ZoeLeBlanc** [CultureAsData-UIUC/is310-fall-2024#10] [Assignments]
Please post your folder link in this discussion. Details about the assignment are available here [https://cultureasdata-uiuc.github.io/is310-fall-2024/materials/creating-curating-humanities-data/06-getting-data-apis#getting-culture-across-apis-homework](https://cultureasdata-uiuc.github.io/is310-fall-2024/materials/creating-curating-humanities-data/06-getting-data-apis#getting-culture-across-apis-homework).

**Comment by haydnelm** [comment]
https://github.com/haydnelm/is310-coding-assignments/tree/main/API-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @haydnelm ! The code works and appreciated learning how to get a STEAM api key. One small thing is that the file does save with a pretty difficult to work with file name since it's hard coding your entire path to your file locally:

<img width="881" alt="Screenshot 2024-11-13 at 9 08 27 PM" src="https://github.com/user-attachments/assets/c10f9238-1bec-406f-8d7e-6c2e87c0a11b">

But it works so happy to give you full marks (though in future, would highly recommend using relative file paths!) 👍

**Comment by ryanpearlman1** [comment]
https://github.com/ryanpearlman1/is310-coding-assignments/tree/main/api_getting_data

**Reply by ZoeLeBlanc** [reply]
Hi @ryanpearlman1 ! Code works great, and impressed you both figured out Google's byzantine API system and stored the output from both APIs in one json file. Well done 👍 !

**Comment by yosefstifanos** [comment]
https://github.com/yosefstifanos/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @yosefstifanos ! Code works great and just want to commend you on saving both API results to a json file--well done 👍 !

**Comment by GloriaLiu930** [comment]
https://github.com/GloriaLiu930/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @GloriaLiu930 ! Code works great but noticed one small issue in your code. Specifically how you're taking the Meow Data and pass it to Europeana. Now that idea is very clever so kudos! 

But there's one small issue with your implementation, which we can see if we print out the search term:

```python
meowfact_data = response_meowfacts.json()
print("MeowFact:", meowfact_data)
search_term = meowfact_data["data"][0]
print("Search Term:", search_term)
```
Which gives us the following output:

```shell
MeowFact: {'data': ['All cats have three sets of long hairs that are sensitive to pressure - whiskers, eyebrows,and the hairs between their paw pads.']}
Search Term: All cats have three sets of long hairs that are sensitive to pressure - whiskers, eyebrows,and the hairs between their paw pads.
```
So for search term we are passing in an incredibly long string that's unlikely to give us hits with the Europeana api. But if we tweak your code slightly to be:

```python
meowfact_data = response_meowfacts.json()
print("MeowFact:", meowfact_data)
search_term = meowfact_data["data"][0].split(" ")[0]
print("Search Term:", search_term)
```
Then the code returns results from both APIs! Anyways this was just a suggestion to help so no need to implement and well done 👍

**Comment by tinayang922** [comment]
https://github.com/tinayang922/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @tinayang922 ! The code works well but just noticed you were only saving the Europeana api results to file. You could save both with just this addition to your code:

```python
combined_data = {
            "Europeana Sports Data": europeana_data["items"],
            "SWAPI Character": the_one_i_want
        }
save_to_json(combined_data)
```

Which gives you a json file with these fields:

<img width="382" alt="Screenshot 2024-11-13 at 9 30 53 PM" src="https://github.com/user-attachments/assets/d6800557-a4d2-497f-8116-5d29fee0b4b5">

Anyways, I'll accept this for full marks since you are saving the data to file but wanted to let you know there was a fairly easy solution save both with your current code. Well done 👍 

--- 
Edited to add that I just realized you've pushed up your Europeana API key with how you've stored the europeana data. I would **strongly recommend** you remove that from code asap to avoid getting banned. I'll give you full marks regardless but please fix this when you can!

**Comment by shoirathi** [comment]
https://github.com/shoirathi/is310-coding-assignments/tree/main/api-getting-data

**Comment by esr27** [comment]
https://github.com/esr27/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @esr27 ! Code works great and especially impressed you got both APIs' data into a `csv` file––well done 👍 !

**Comment by OctaneCodes** [comment]
https://github.com/OctaneCodes/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @OctaneCodes ! Code works great and impressed you got both outputs into a json, well done 👍 !

**Comment by Cindy1226** [comment]
https://github.com/Cindy1226/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @Cindy1226 ! Your code works great but I'm concerned for a few issues: first, it looks like you may have pushed up your Marvel public api key and you pushed up your Europeana api key in your `marvel_europeana_data` file. Second, your `marvel_europeana_data` file is missing a file extension. You seem to have made the choice intentionally based on the comments in your code but not sure why?

<img width="343" alt="Screenshot 2024-11-17 at 2 22 10 PM" src="https://github.com/user-attachments/assets/9e4d15f8-7341-47b2-999e-09a8d3282183">

Anyways, happy to give you full marks, but would highly recommend fixing the api keys issues before you get banned by either API. Otherwise, well done 👍

**Comment by SamuelSokolovsky** [comment]
https://github.com/SamuelSokolovsky/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @SamuelSokolovsky ! Code is working great and impressed with your choice of API. However, you did push up your Europeana api key so would recommend fixing that before you get banned from that API. Otherwise well done 👍

**Comment by adg77** [comment]
https://github.com/adg77/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Great work @adg77 ! Liked the choice of API 👍

**Comment by kohtas2** [comment]
https://github.com/kohtas2/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @kohtas2 ! Code works great but you accidentally pushed up your Europeana api key in the json data, so would highly recommend removing that. Otherwise, well done 👍

**Comment by lhayedimarucut** [comment]
https://github.com/lhayedimarucut/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @lhayedimarucut ! Code works great and impressed at you combining both data in the json file, well done 👍 !

**Comment by LucasWashor** [comment]
https://github.com/LucasWashor/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @LucasWashor  ! Code works great but you accidentally pushed up your Europeana api key in the json data, so would highly recommend removing that. Otherwise, well done 👍

**Comment by JBrown-0** [comment]
https://github.com/JBrown-0/is310-coding-assignments/tree/main/python-libraries/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @JBrown-0 ! Very impressed that you used such a difficult API to setup (I'll be honest this one took me quite some time to get working locally) and the code seems to be working. One issue though is that you pushed up your Europeana api key in your json file, so would highly recommend removing that before you get banned. Otherwise very well done 👍

**Comment by jasonmorris8** [comment]
https://github.com/jasonmorris8/is310-coding-assignments/tree/main/api-getting-data

**Comment by henRotunno** [comment]
https://github.com/henRotunno/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @henRotunno! Your code is very close but noticing two issues. The first isn't huge but you pushed up your Europeana api key in your code so would highly recommend fixing that.

The second is that currently the code does not complete successfully. Specifically I get the following error:

```shell
SWAPI Data: {
  "name": "Luke Skywalker",
  "height": "172",
  "mass": "77",
  "hair_color": "blond",
  "skin_color": "fair",
  "eye_color": "blue",
  "birth_year": "19BBY",
  "gender": "male",
  "homeworld": "https://swapi.dev/api/planets/1/",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/2/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/6/"
  ],
  "species": [],
  "vehicles": [
    "https://swapi.dev/api/vehicles/14/",
    "https://swapi.dev/api/vehicles/30/"
  ],
  "starships": [
    "https://swapi.dev/api/starships/12/",
    "https://swapi.dev/api/starships/22/"
  ],
  "created": "2014-12-09T13:50:51.644000Z",
  "edited": "2014-12-20T21:17:56.891000Z",
  "url": "https://swapi.dev/api/people/1/"
}
Traceback (most recent call last):
  File "/Users/zleblanc/teaching/is310-Fall-2024/new_homework_assignments/henryjr2/is310-coding-assignments/api-getting-data/api-sc.py", line 24, in <module>
    scan = star_dat['Vader']
KeyError: 'Vader'
```
As you can tell, it's throwing an error because the key `Vader` does not exist in the Star Wars data. So you need to fix that issue and then I'm happy to give you full marks. Also feel free to reach out if you get stuck or have questions!

**Comment by averyWebster** [comment]
https://github.com/averyWebster/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @averyWebster ! This code looks great but one issue. You pushed up your Europeana API key. I would **strongly recommend** removing that line of code asap. Apologies for not catching that sooner but otherwise well done 👍 

Edited to add, I just noticed you had a `.gitignore` inside your `api-getting-data` folder. You really only need one usually at the root of your folder where you do the `git init` command. So in this case in your `is310-coding-assignments` folder. It looks like you already have one there but might need to copy over the virtual environment name. Once you do that, you should be able to delete the one in your `api-getting-data` folder. The reason you don't want multiples is that it can cause issues with git knowing which one to listen to or not. If you have any issues or questions, please don't hesitate to reach out!

**Comment by ahmadsawwan** [comment]
https://github.com/ahmadsawwan/ahmadsawwan-is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @ahmadsawwan ! Very impressed with this homework and was able to get both APIs working. You just have one small issue. The way you've done the file paths throws and error 😅 

<img width="956" alt="Screenshot 2024-11-11 at 10 43 27 PM" src="https://github.com/user-attachments/assets/24a9f342-6444-4bc0-abb4-27f209919c2c">

I believe if you remove the directory part of where you're saving the file it should work without an error. Essentially the issue is I'm running your code from that directory so the computer can't find a directory I'm already in to save a file into. Hope that makes sense and happy to answer any questions you have! Mostly you're very close to being done so if you could just fix that aspect then you'll get full marks 👍

**Reply by ahmadsawwan** [reply]
I think this should be fixed now

**Reply by ZoeLeBlanc** [reply]
Very close but still doesn't work 😅 . Again, try running this script from inside that `api-getting-data` directory (so cd `api-getting-data` then `python getting_culture.py`) and you'll see why it is not working.

**Reply by ahmadsawwan** [reply]
Ok this time I for sure fixed it. it works when I run the code from inside api-getting-data!

**Reply by ZoeLeBlanc** [reply]
It works! Great fix 👍

**Comment by elee33** [comment]
https://github.com/elee33/is310-coding-assignments/tree/main/the_one_api_project

**Reply by ZoeLeBlanc** [reply]
Hi @elee33 ! Your code is working great but noticed a few issues. First you pushed your Europeana api key both in your code and data. Second, you also have a `.gitignore` file in your assignment folder, which is unnecessary since you have one in your root directory. Finally, you named this folder `the_one_api_project` but you aren't using that api so would recommend cleaning that up for clarity. Otherwise good work 👍

**Comment by officalcharles** [comment]
https://github.com/officalcharles/is310-coding-assignments/tree/master/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @officalcharles ! Code works great and just wanted to commend you on saving both datasets in the same json file. That's very clever so well done 👍

**Comment by AlanK3** [comment]
https://github.com/AlanK3/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @AlanK3 ! The code works well though wanted to flag two tweaks I had to make in case you're having issues (though this might be an issue just for my laptop). 

For the MTD API, I needed to append my key directly as a query rather than passing it as a param:

```python
f"https://developer.mtd.org/api/v2.2/json/getvehicles?key={MTD_API_KEY}"
```
For Europeana API, I needed to remove the quotations around `EUROPEANA_API_KEY` when passing the params:

```python
 os.environ['EUROPEANA_API_KEY'] = EUROPEANA_API_KEY
    base_url = "https://api.europeana.eu/record/v2/search.json"
    params = {
        'wskey': EUROPEANA_API_KEY,
        'query': query,
        'rows': 5
    }
   ```
   
   Hope this helps and well done again 👍

**Comment by theajk2** [comment]
https://github.com/theajk2/is310-coding-assignments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @theajk2 ! Your code is technically working in that your script doesn't throw an error, but that's because nothing is running currently.

<img width="566" alt="Screenshot 2024-11-18 at 11 56 43 AM" src="https://github.com/user-attachments/assets/605fb069-3505-4d1e-b253-6bf7b0e2a828">

Specifically, as you can see above, you are not calling your `main()` function. I imagine you were trying to use this syntax to run the code but notice the issues with your approach:

```python
if __name__ == "__main__":
        return
```

First this block of code is indented under the `main` function so it will only work if `main()` is called. But then even if you do the indentation properly, you need to have the `main()` function call with that code. So it should be:

```python
if __name__ == "__main__":
        main()
```

When I run that code, I get a lot of errors, so seems like you have some issues to fix. Happy to help if you have questions, so please don't hesitate to reach out!

**Comment by Rebe-Rod** [comment]
https://github.com/Rebe-Rod/is310-coding-assingments/tree/main/api-getting-data

**Reply by ZoeLeBlanc** [reply]
Hi @Rebe-Rod ! I tried running your code but found a number of issues, so wondering if it worked for you? First, you aren't actually calling your `main()` function and instead the code immediately tries to run this since it is the final lines of your script and the only ones that aren't a function:

```python
with open ('quotes.json', 'w') as f:
    json.dump(get_quotes, f, indent=2)
 ```

This throws the following error:

```shell
  File "/Users/zleblanc/teaching/is310-Fall-2024/new_homework_assignments/rrodr23/is310-coding-assignments/api-getting-data/getting_culture.py", line 38, in <module>
    json.dump(get_quotes, f, indent=2)
  File "/opt/homebrew/Cellar/python@3.9/3.9.20/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/opt/homebrew/Cellar/python@3.9/3.9.20/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/opt/homebrew/Cellar/python@3.9/3.9.20/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type function is not JSON serializable
```

Which is telling us that you cannot write a function (in this case your `get_quotes` function) to json. Happy to help if you're stuck so don't hesitate to reach out if you have questions, but currently this code needs some work to actually complete successfully.

**Comment by frangara** [comment]
https://github.com/frangara/is310-coding-assignments/tree/main/api-getting-data

Thank you for your help on this assignment. It was really cool to see how APIs work and how I can find one and relate it to the europeana library

**Comment by AG-Samaniego** [comment]
https://github.com/AG-Samaniego/is310-coding-assignments/tree/main/api-getting-data
(Ran into issues for a bit so I wasn't able to submit on Saturday as I had originally hoped; also don't mind the code comments to myself lol)

**Comment by chuzesta** [comment]
https://github.com/chuzesta/is310-coding-assignments/tree/main/api-getting-data
finished on time but forgot to submit on here.

**Comment by divyaj5** [comment]
https://github.com/divyaj5/is310-coding-assignments/tree/master/api-getting-data

**Comment by yukthas2** [comment]
https://github.com/yukthas2/is310-coding-assignments/tree/main/api-getting-data
